Add "Installing PHP Locally (for Symfony CLI)" post (#52)
knpuniversity/blog: install-php-locally.md

# Installing PHP Locally (for Symfony CLI)

The [Symfony CLI](https://symfony.com/doc/current/setup/symfony_cli.html) is awesome for developing Symfony
applications locally! It enables you to run a local web server, integrates with Docker containers for
services like a database, and provides many other features to streamline your development workflow.
It's mostly batteries included, but it does require PHP to be installed locally.

We use it here at SymfonyCasts for almost all our courses. Let's look at how to install PHP on different
operating systems. We'll also make sure extensions for the most common databases (MySQL, SQLite, PostgreSQL)
are installed.

> Make sure you have the [Symfony CLI installed](https://symfony.com/download) before getting started!

Author: > I think you can because you could use Docker to manage your PHP

I'm going to sort of ignore this for the purpose of this post. The Symfony and SymfonyCasts recommended way is to install PHP locally. I'll adjust the sentence to not say "required", but "recommended". Or maybe, it just "needs" php... and we recommend the following methods to install on your OS.

## macOS (Homebrew)

First, ensure you have [Homebrew](https://brew.sh) installed.

> Next, install PHP:

Reviewer: Sounds like we can add a header to match the header about extensions below:

```suggestion
### Homebrew PHP

Next, install PHP:
```

```terminal
brew install php
```

This will install the latest stable version. Ensure it's working by running:

```terminal
symfony check:requirements

> ```

Reviewer: I would move this to the end, out of the OS-related section. Because no matter how you install the PHP, this command will be the same for any OS.

Reviewer 2: On the other hand, that's what you always do after installing something, check if it works

Author: I'm going to move to a final section (that all OS-specific sections will direct to)

PHP should be ready to go with common extensions enabled. Also, the MySQL, SQLite, and PostgreSQL extensions
should be installed by default. Double check by running:

```terminal
php -m | grep pdo
```

### Homebrew PHP Extensions

> If you need additional extensions, not included, use [pecl](https://pecl.php.net/), which was installed with PHP.

Reviewer: ```suggestion
If you need additional extensions, not included, use [PECL](https://pecl.php.net/) (PHP Extension Community Library), which was installed with PHP.
```

For instance, to install the Redis extension, run:

```terminal
pecl install redis
```

## Ubuntu/Debian Linux (apt)

In Ubuntu and Debian, you can install PHP using the `apt` package manager. Your default package repository probably
won't have the latest PHP version. The 3rd-party `ondrej/php` repository is a popular choice that provides
the latest PHP versions and extensions. Add it to your system by running:

```terminal
sudo add-apt-repository ppa:ondrej/php
sudo apt update
```

To install PHP 8.4 and some common extensions, run:

```terminal
sudo apt install php8.4 php8.4-cli php8.4-common php8.4-curl php8.4-mbstring php8.4-xml php8.4-zip php8.4-gd php8.4-intl php8.4-bcmath 
```

To install the common database extensions, run:

```terminal
sudo apt install php8.4-mysql php8.4-pgsql php8.4-sqlite3
```

***NOTE
Replace `8.4` with the version you want to install. Example: `php8.5 php8.5-cli ...` for PHP 8.5.
***

Ensure all looks good by running:

```terminal
symfony check:requirements
```

### APT PHP Extensions

> For additional PHP extensions, you don't need to use `pecl`, these are also available via `apt`. For example, to install

Reviewer: I think some people may skip reading macOS section and would be confused about it:
```suggestion
For additional PHP extensions, you don't need to use `pecl` that we mentioned in the macOS section. These are also available via `apt`. For example, to install
```

```terminal
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

### Homebrew PHP

Now, install PHP:

```terminal
brew install php
```

***TIP
This will install the latest stable version of PHP. If you want a specific version, check
[the Homebrew PHP formula](https://formulae.brew.sh/formula/php). Under "Other versions"
you'll find available alternate versions. For example, to install PHP 8.3, run:

```terminal
brew install php@8.3
```
***

### Homebrew PHP Extensions

Homebrew's PHP formula includes many extensions out of the box (including common database extensions).
If you need additional extensions, use [PECL](https://pecl.php.net/) (PHP Extension Community
Library), which was installed with PHP. For instance, to install the Redis extension, run:

```terminal
pecl install redis
```

### Homebrew Symfony CLI

There are several methods to [install the Symfony CLI](https://symfony.com/download), since we already
have Homebrew installed, let's use it to install the Symfony CLI:

```terminal
brew install symfony-cli/tap/symfony-cli
```

> You should be ready to go on your Mac! [Verify the installation below](#verify-phpsymfony-cli-installation).

Reviewer: IMO, better to link to the official website on it too. But I see your point, if we really want to make this more opinionated - I'm fine with it

sudo apt update
```

To install PHP 8.4 and some common extensions, run:

```terminal
sudo apt install php8.4 php8.4-cli php8.4-common php8.4-curl php8.4-mbstring php8.4-xml php8.4-zip php8.4-gd php8.4-intl php8.4-bcmath 
```

To install the common database extensions, run:

```terminal
sudo apt install php8.4-mysql php8.4-pgsql php8.4-sqlite3
```

***NOTE
Replace `8.4` with the version you want to install. For example: `php8.3 php8.3-cli ...` for PHP 8.3.
***

### APT PHP Extensions

For additional PHP extensions, these are also available via `apt`. For example, to install
the Redis extension, run:

```terminal
sudo apt install php8.4-redis
```

### APT Symfony CLI

There are several methods to [install the Symfony CLI](https://symfony.com/download), since we used
`apt` to install PHP, let's use it to install the Symfony CLI as well:

```terminal
curl -1sLf 'https://dl.cloudsmith.io/public/symfony/stable/setup.deb.sh' | sudo -E bash
sudo apt install symfony-cli
```

You should be ready to go! [Verify the installation below](#verify-phpsymfony-cli-installation).

## Windows (WSL)

Running PHP natively on Windows (as an `exe`) is *possible*, but can be tricky. I recommend using the

> "Windows Subsystem for Linux" (WSL) to run Ubuntu. It isn't in a container or anything, it runs

Reviewer: Maybe this?
```suggestion
"Windows Subsystem for Linux" (WSL) to run Ubuntu. It isn't in a container or or other virtualization tool, it runs
```

*directly* on Windows! Follow the [WSL installation guide](https://learn.microsoft.com/en-us/windows/wsl/install)
to get setup with WSL and Ubuntu.

Now, follow the [Ubuntu/Debian Linux (apt)](#ubuntudebian-linux-apt) section above to get PHP installed!

## Verify PHP/Symfony CLI Installation

Check your PHP version by running:

```terminal
php -v
```

Check your Symfony CLI version by running:

```terminal
symfony version
```

You can run PHP *through* the Symfony CLI. Ensure the following matches your PHP version above:

```terminal
symfony php -v
```

For the common database extensions, ensure they are all installed by running:

```terminal
symfony php -m | grep pdo
```

You should see the following output:

```
pdo_mysql # MySQL extension
pdo_pgsql # PostgreSQL extension
pdo_sqlite # SQLite extension
```

Finally, to ensure that you're ready to go with Symfony, run:

```terminal
symfony check:requirements
```

> 

Reviewer: How about adding a note that you can use this command in the root directory of your Symfony project to check project-specified requirements?
